stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+
+        let coordinator_config = CoordinatorConfig {
+            threshold,
+            dkg_threshold,
+            num_signers,
+            num_keys,
+            message_private_key: signer_config.ecdsa_private_key,
+            dkg_public_timeout: signer_config.dkg_public_timeout,
+            dkg_private_timeout: signer_config.dkg_private_timeout,
+            dkg_end_timeout: signer_config.dkg_end_timeout,
+            nonce_timeout: signer_config.nonce_timeout,
+            sign_timeout: signer_config.sign_timeout,
+            signer_key_ids: signer_config.registered_signers.coordinator_key_ids,
+            signer_public_keys: signer_config.registered_signers.signer_public_keys,
+        };
+
+        let coordinator = FireCoordinator::new(coordinator_config);
+        let signing_round = WSTSSigner::new(
+            threshold,
+            num_signers,
+            num_keys,
+            signer_config.signer_id,
+            signer_config.key_ids,
+            signer_config.ecdsa_private_key,
+            signer_config.registered_signers.public_keys.clone(),
+        );
+        let coordinator_selector =
+            CoordinatorSelector::from(signer_config.registered_signers.public_keys);
+
+        debug!(
+            "Signer #{}: initial coordinator is signer {}",
+            signer_config.signer_id,
+            coordinator_selector.get_coordinator().0
+        );
+
+        Self {
+            coordinator,
+            signing_round,
+            state: State::Idle,
+            blocks: HashMap::new(),
+            commands: VecDeque::new(),
+            stackerdb,
+            mainnet: signer_config.mainnet,
+            signer_id: signer_config.signer_id,
+            signer_ids: signer_config
+                .registered_signers
+                .signer_ids
+                .values()
+                .copied()
+                .collect(),
+            signer_slot_ids: signer_config.registered_signers.signer_slot_ids,
+            next_signer_ids: vec![],
+            next_signer_slot_ids: HashMap::new(),
+            reward_cycle: signer_config.reward_cycle,
+            tx_fee_ustx: signer_config.tx_fee_ustx,
+            coordinator_selector,
+        }
+    }
+}
+
+impl Signer {
+    /// Finish an operation and update the coordinator selector accordingly
+    fn finish_operation(&mut self) {
+        self.state = State::Idle;
+        self.coordinator_selector.last_message_time = None;
+    }
+
+    /// Update operation
+    fn update_operation(&mut self) {
+        self.state = State::OperationInProgress;
+        self.coordinator_selector.last_message_time = Some(Instant::now());
+    }
+
+    /// Execute the given command and update state accordingly
+    fn execute_command(&mut self, stacks_client: &StacksClient, command: &Command) {
+        match command {
+            Command::Dkg => {
+                //TODO: check if we already have an aggregate key stored in the contract.
```

Comment:
Please open a ticket for this if there isn't one already.

Reply:
Ah this is easy to add. Will do right now.